pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.ast
+
+import com.google.summit.ast.*
+import com.google.summit.ast.declaration.*
+import com.google.summit.ast.modifier.KeywordModifier
+import com.google.summit.ast.modifier.Modifier
+import net.sourceforge.pmd.annotation.InternalApi
+import net.sourceforge.pmd.lang.apex.ApexParserOptions
+import net.sourceforge.pmd.lang.ast.ParseException
+import net.sourceforge.pmd.lang.ast.SourceCodePositioner
+
+@Deprecated("internal")
+@InternalApi
+@Suppress("DEPRECATION")
+class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptions) {
+    private val sourceCodePositioner = SourceCodePositioner(sourceCode)
+
+    /** Builds and returns an [ApexNode] AST corresponding to the given [root] node. */
+    fun build(root: CompilationUnit): ApexRootNode<TypeDeclaration> =
+        visit(root, parent = null) as? ApexRootNode<TypeDeclaration>
+            ?: throw ParseException("Unable to build tree")
+
+    /**
+     * Visits [node] and returns an [ApexNode] wrapper.
+     *
+     * Sets the parent of the resulting [ApexNode] to [parent], if it's not `null`.
+     */
+    private fun visit(node: Node?, parent: ApexNode<*>?): AbstractApexNode<*>? {
+        val wrapper: AbstractApexNode<*>? =
+            when (node) {
+                null -> null
+                is CompilationUnit -> visit(node.typeDeclaration, parent)
+                is TypeDeclaration -> visitTypeDeclaration(node)
+                is Identifier,
+                is KeywordModifier,
+                is TypeRef -> null
+                else -> {
+                    println("No adapter exists for type ${node::class.qualifiedName}")
+                    // TODO(b/239648780): temporary print
+                    null
+                }
+            }
+
+        wrapper?.setParent(parent)
+        wrapper?.handleSourceCode(sourceCode)
+        wrapper?.calculateLineNumbers(sourceCodePositioner)
+        return wrapper
+    }
+
+    /** Calls [visit] on each of [nodes]. */
+    private fun visit(nodes: List<Node>, parent: ApexNode<*>?) = nodes.forEach { visit(it, parent) }
+
+    /** Calls [visit] on each [child][Node.getChildren] of [node]. */
+    private fun visitAll(node: Node, parent: ApexNode<*>?) =
+        node.getChildren().forEach { visit(it, parent) }
+
+    /**
+     * Calls [visit] on each [child][Node.getChildren] of [node] that isn't already [visited].
+     *
+     * [visited] is a predicate that, given a child node `child`, should return `true` when `child` is
+     * equal to a property of [node] that was already visited and shouldn't be visited again.
+     */
+    private fun visitRest(node: Node, parent: ApexNode<*>?, visited: (Node) -> Boolean) =
```

Comment:
thoughts on (possibly combining with the above into):

```
private fun alwaysTrue(n: Node): Boolean = true

// defaults to visiting all children, but accepts an `onlyChildren` predicate to filter
private fun visitChildren(node: Node, parent: ApexNode<*>?, onlyChildren: (Node) -> Boolean = ::alwaysTrue) ...
```

Reply:
This is a good way to combine the two functions. However, do you think the predicate should leave the matching children in or out?

Filtering out, like `visitRest` does currently, has the benefit of not having to name the remaining children explicitly, and only repeating what was handled specially, e.g.
```kotlin
visitModifiers(node.modifiers, parent = this)
visitChildren(node, parent = this, exclude = { it in node.modifiers })
```
vs.
```kotlin
visitModifiers(node.modifiers, parent = this)
visitChildren(node, parent = this, include = { /* unique to each declaration, despite performing same operation each time */ })
```

Reply:
I don't have a strong opinion, whichever you think works best. (But don't the two functions only differ by an extra`!` ?)

Reply:
Sorry, somehow I missed the fact that you can negate a boolean. In that case, I only have a slight preference for `exclude`, since the predicate lambda will be a bit more readable. Thank you